lib/llog/api.go

```diff
@@ -0,0 +1,201 @@
+package llog
+
+import (
+	"io"
+)
+
+// Level indicates logging level. The higher the value, the less important.
+// Any non-negative integer is always less important than a named constant.
+type Level int
+
+const (
+	// Least important
+	Info Level = -1 - iota
+	Warn
+	// Most important
+	Err
+)
+
+// Set sets this level from string equivalent. s must be "Err",
+// "Warn", "Info", "0", "1", "2", ... or else Set returns an error. Set
+// ignores the case in s, so passing "ERR", "Err", or "err" as s has the same
+// effect.
+func (l *Level) Set(s string) error {
+	// TODO
+	return nil
+}
+
+// String returns the string representation of this level in all caps.
+func (l *Level) String() string {
+	// TODO
+	return ""
+}
+
+// FlushWriter allows the logging of fatal and panic messages to notify the
+// underlying writer that the process is terminating.
+type FlushWriter interface {
+	io.Writer
+	// Logging a fatal or panic message calls this just before terminating
+	// the process.
+	Flush() error
+}
+
+// A Sink synchronizes the writing of log messsges to a particular Writer.
+// A Sink instance is safe to use with multiple goroutines. A client will
+// create one Sink to control access to a single Writer and then create one
+// or more Logger instances, each with possibly different configurations,
+// to write to that same Sink.
+type Sink struct {
```

Comment:
Yes, making the writer deal with concurrent calls would obviate the need for a Sink layer, but consider this: logger instances in the log package already synchronise writes to the underlying writer. If you use the built in go log package, you don't have to worry about making your underlying writer thread-safe. See the docs here: https://golang.org/pkg/log/#Logger

Since typically go programs create one Logger instance, making the logger instance synchronise writes was a good design choice. However, because we will be creating a separate logger for each subsystm (to facilitate filtering by subsystem) that design won't work for us. We have three options:

1. Force the client of our API to ensure their writers are threadsafe: While this obviates the need for the sink layer, it forces the client of our API to do extra work since most writers are not thread safe. It would be easy for our clients to make mistakes here. At the minimum, we'd want to provide a thread-safe adapter for io.Writer.

2. Use the Sink abstraction. While this doesn't pass the buck, it does add a layer of complexity. The advantage here is that caller can't make mistakes around synchronising writes.

3. Use a global mutex to synchronise writes. With this option, you would create multiple loggers on top of the same writer, but they would use a global mutex to coordinate writing. This provides the best of both worlds: client doesn't have to worry about thread safety and simple API. The disadvantage with this option is extra contention in applications with multiple logs. Logging messages would contend for the global mutex even if they are writing to different writers.

Of all of these choices, I am inclined to go with 3) as I believe that most applications will use a single writer. 3) provides the simplest and most robust API 3) will still work with multiple writers but with extra contention.